paddle/fluid/pir/transforms/dead_code_elimination_pass.cc

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/pir/transforms/dead_code_elimination_pass.h"
+
+#include "paddle/fluid/pir/dialect/operator/ir/pd_op.h"
+#include "paddle/pir/core/builtin_op.h"
+#include "paddle/pir/core/program.h"
+#include "paddle/pir/pass/pass.h"
+#include "paddle/pir/pass/pass_registry.h"
+#include "paddle/pir/pattern_rewrite/frozen_rewrite_pattern_set.h"
+#include "paddle/pir/pattern_rewrite/pattern_match.h"
+#include "paddle/pir/pattern_rewrite/pattern_rewrite_driver.h"
+
+namespace {
+
+class DeadCodeEliminationPattern : public pir::RewritePattern {
+ public:
+  DeadCodeEliminationPattern(
+      pir::IrContext* context,
+      pir::PatternBenefit benefit = 1,
+      const std::vector<std::string>& generated_names = {})
+      : RewritePattern(MatchAnyOpTypeTag(), benefit, context, generated_names) {
+  }
+
+  bool Match(pir::Operation* op) const override {
+    if (op->isa<paddle::dialect::FetchOp>() ||
+        op->isa<paddle::dialect::ShadowOutputOp>())
+      return false;
+
+    return op->use_empty();
+  }
+
+  void Rewrite(pir::Operation* op,
+               pir::PatternRewriter& rewriter) const override {  // NOLINT
+    if (op->dyn_cast<pir::GetParameterOp>()) {
+      // Delete parameter from program.
+      pir::GetParameterOp get_parameter_op =
+          op->dyn_cast<pir::GetParameterOp>();
+      get_parameter_op->GetParentProgram()->parameters().erase(
+          get_parameter_op->attributes()
+              .at(get_parameter_op.attributes_name[0])
+              .dyn_cast<pir::StrAttribute>()
+              .AsString());
+    }
+    rewriter.EraseOp(op);
+  }
+};
+
+class DeadCodeEliminationPass : public pir::Pass {
+ public:
+  DeadCodeEliminationPass() : pir::Pass("dead_code_elimination_pass", 0) {}
+
+  bool Initialize(pir::IrContext* context) override {
+    pir::RewritePatternSet ps(context);
+    ps.Add<DeadCodeEliminationPattern>(context);
+    patterns_ = pir::FrozenRewritePatternSet(std::move(ps));
+    return true;
+  }
+
+  void Run(pir::Operation* op) override {
+    pir::GreedyRewriteConfig cfg;
+    cfg.use_top_down_traversal = true;
+    cfg.max_iterations = 10;
```

Comment:
Why this magic number？

Reply:
It controls the number of iterations and patterns during full graph matching.